### ISSUE TITLE
IllegalMonitorStateException in Connection#doShutDown

### ISSUE THREAD

**Issue by tau3** [bisq-network/bisq#1315]
Hello! 

I've found a comment about `IllegalMonitorStateException` in `io.bisq.network.p2p.network.Connection#doShutDown`. 

I think that the problem is:
1. Some code invokes the method with `CloseConnectionReason` which requires to send message about closing;
2. The message is sent in a separate thread (see `Connection#shutDown(CloseConnectionReason, Runnable`)) in current implementation;
3. `shutDown()` method invokes `doShutDown()` through `UserThread`. But the current executor of `UserThread` (`MoreExecutors#directExecutor`) just runs all jobs in the same thread that has called `UserThread`;
4. So `doShutdown()` method is actualy invoked in a thread that was created in step 2 (a _different_ thread);
5. The thread is definitiely not the owner of `Connection#protoOutputStreamLock` which is attempted to be unlocked. That is why the exception occurs IMO.

tl;dr ReentrantLock is locked in one thread and unlocked in another.

**Comment by ManfredKarrer** [comment]
Good analysis! Thanks!

The shutdown is called from user threads as well as from the connection itself. We might need to refactor that to handle those 2 different thread context differently. Feel free to make a suggestion!

**Comment by tau3** [comment]
TBH, i'm not actually sure that i completely understand locks flow in `Connection` class, so my suggestion is quite straightforward - can't we just execute shutdown in the same thread in case of sending close message? Would it cause any problems?

**Comment by ManfredKarrer** [comment]
Yes that should work. Just the shutdown calls from outside need to be considered. Feel free to refactor the locking handling there. The Connection class could need some love ;-).

**Comment by ManfredKarrer** [comment]
Merged in https://github.com/bisq-network/bisq/pull/1318

Btw: @tau3 Feel free to make a compensation request for all your past contributions!
